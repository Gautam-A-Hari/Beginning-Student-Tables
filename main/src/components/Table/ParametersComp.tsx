import React from "react";
import { peekKey, takeKey } from "../App";
import { RemButton } from "../RemButton";
import { Environment, Yellow } from "../../global-definitions";
import { ExampleArray, Parameter, ParameterArray, isParameterNonEmpty } from "../../input-definitions";
import { ValidatedInput } from "../ValidatedInput";
import { nameRE } from "../../parser";

interface Props {
    params: ParameterArray
    tableNames: Array<string | Yellow | undefined>
    globalEnv: Environment
    examples: ExampleArray
    disabled: boolean

    paramsExamplesChange: (params: ParameterArray, examples: ExampleArray) => void

}
function ParametersComp(props: Props) {
    function validParam(text: (string | undefined | Yellow), modParam: ({} | Parameter)) {
        function inEnv(name: (string | undefined | Yellow), env: Environment) {
            return env.reduce((acc, variable) => {
                if (acc) {
                    return true;
                }

                return variable.name === name;

            }, false);
        }

        // These are not technically Variables, see note above
        let paramVars = props.params.filter((param) => param !== modParam).map((param) => ({ name: param.name, binding: null }));
        let tableVars = props.tableNames.map((name) => ({ name: name, binding: null }));
        let env: Environment = [...props.globalEnv, ...tableVars, ...paramVars];

        return typeof text === "string" ? nameRE.test(text) && !inEnv(text, env) : false;
    }

    function remParam(deadParam: Parameter) {
        const deadIndex = props.params.indexOf(deadParam);
        const aliveParams = props.params.filter((param) => param !== deadParam);

        // need to maintain #inputs = #params
        const modExamples = props.examples.map((example => ({
            ...example,
            inputs: example.inputs.filter((_, i) => i !== deadIndex)
        })));

        props.paramsExamplesChange(aliveParams, modExamples);
    }

    // String, Number -> Side Effect
    // changes the name of the given parameter
    // if that parameter doesn't exist yet, it makes it
    function paramChange(newParam: Parameter, modParam: ({} | Parameter)): void {

        let alteredParams: ParameterArray, alteredExamples: ExampleArray;
        // check if the param exists in params
        if ((isParameterNonEmpty(modParam) && props.params.indexOf(modParam) === -1) || !isParameterNonEmpty(modParam)) {
            alteredParams = [...props.params, newParam];
            alteredExamples = props.examples.map((example) => ({
                ...example,
                inputs: [...example.inputs, { prog: { raw: '', validated: { yellow: 'yellow' } }, key: takeKey() }]
            }));
        } else {
            alteredParams = props.params.map((param) => param === modParam ? newParam : param);
            alteredExamples = props.examples;
        }
        return props.paramsExamplesChange(alteredParams, alteredExamples);
    }

    const reals = props.params.map((param) => (
        <th key={param.key} >
            <div className='flex_horiz'>
                <ValidatedInput
                    dummy={false}
                    placeholder='Parameter'
                    text={props.disabled ? param.name : undefined}
                    rawText={""}
                    isValid={(text: (Yellow | string)) => validParam(text, param)}
                    onValid={(text: (Yellow | string)) => paramChange({
                        ...param,
                        name: text
                    },
                        param)}
                    onEmpty={() => paramChange({
                        ...param,
                        name: { yellow: 'yellow' }
                    },
                        param)}
                />
                <RemButton
                    title='Remove this parameter'
                    onClick={props.disabled ? undefined : (() => remParam(param))}
                />
            </div>
        </th>
    ));

    const dummy = (
        <th key={peekKey()}>
            <div className='flex_horiz'>
                <ValidatedInput
                    dummy={true}
                    placeholder='Parameter'
                    text={props.disabled ? '' : undefined}
                    rawText={''}
                    isValid={(text) => validParam(text, {})}
                    onValid={(text) => paramChange({
                        name: text,
                        key: takeKey()
                    },
                        {})}
                    onEmpty={() => null}
                />
            </div>
        </th>
    );

    return (
        <React.Fragment>
            <th>{/* empty cell to align with example RemButtons */}</th>
            <React.Fragment>
                {[...reals, dummy]}
            </React.Fragment>
        </React.Fragment>
    );
}

export { ParametersComp };