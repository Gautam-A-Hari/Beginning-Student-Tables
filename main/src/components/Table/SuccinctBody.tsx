import React, { RefObject } from 'react';
import { DragDropContext, Draggable, Droppable } from 'react-beautiful-dnd';
import { peekKey, takeKey } from '../App';
import { RemButton } from "../RemButton";
import { Environment } from '../../global-definitions';
import { Example, ExampleArray, Formula, FormulaArray, InputArray, isBooleanFormula, isExampleNonEmpty, NameInput, OutputArray, ProgramInput } from '../../input-definitions';
import { Inputs } from './Inputs';
import { Want } from './Want';
import { Outputs } from './Outputs';

interface Props {
    disabled: boolean
    globalEnv: Environment
    examples: ExampleArray
    formulas: FormulaArray
    tableIndex?: number
    paramNames: Array<NameInput>

    formulasChange: (formulas: FormulaArray) => void
    examplesFormulasChange: (alteredExamples: ExampleArray, alteredForms: FormulaArray) => void
    handleOnDrag: (exampleArr: ExampleArray, tabIndex: number, newExample?: boolean) => void
}

function SuccinctBody(props: Props) {
    function remExample(deadExample: Example) {
        const deadIndex = props.examples.indexOf(deadExample);
        // Formula -> Formula
        // removes the output at deadIndex from the given formula and all of its children (if it has any) so stuff works
        function removeOutputFromFormula(formula: Formula) {
            let outputs = formula.outputs.filter((_: any, i: number) => i !== deadIndex);

            if (isBooleanFormula(formula)) {
                const thenChildren: FormulaArray = formula.thenChildren.map(removeOutputFromFormula);
                return {
                    ...formula,
                    outputs,
                    thenChildren
                };
            } else {
                return {
                    ...formula,
                    outputs
                };
            }
        }


        const aliveExamples = props.examples.filter((example) => example !== deadExample);
        const alteredForms = props.formulas.map(removeOutputFromFormula);
        props.examplesFormulasChange(aliveExamples, alteredForms);
    }

    // split into function that makes a new and function that handles not exist exist case
    function exampleChange(newExample: Example, oldExample: (Example | {})) {
        const exists = isExampleNonEmpty(oldExample) ? props.examples.indexOf(oldExample) !== -1 : false;

        // Formula -> Formula
        // adds an init output to the given formula and all of its children (if it has any) so stuff works
        function addAnotherOutputToFormula(formula: Formula): Formula {
            let outputs: OutputArray = [...formula.outputs, { yellow: 'yellow' }];

            if (isBooleanFormula(formula)) {
                const thenChildren = formula.thenChildren.map(addAnotherOutputToFormula);
                return {
                    ...formula,
                    outputs,
                    thenChildren
                };
            } else {
                return {
                    ...formula,
                    outputs
                };
            }
        }

        let alteredExamples, alteredForms;
        if (exists) {
            alteredExamples = props.examples.map((example) => example === oldExample ? newExample : example);
            alteredForms = props.formulas;
        } else {
            alteredExamples = [...props.examples, newExample];
            alteredForms = props.formulas.map(addAnotherOutputToFormula);
        }

        props.examplesFormulasChange(alteredExamples, alteredForms);
        return; // this doesn't actually do anything
    }

    // handleOnRowDrag : DragObject -> 
    // passes an updated Example order to handleOnDrag
    // result is HTMLInputObject type thing
    function handleOnRowDrag(result: any): any {
        if (!result.destination || result.source.index === result.destination.index) {
            return;
        }
        const sourceIndex = result.source.index;
        const destinationIndex = result.destination.index;
        const examplesList: ExampleArray = Array.from(props.examples);
        let newYellowExample: Example;

        if (sourceIndex === examplesList.length) {
            newYellowExample = { inputs: [{ prog: { raw: '', validated: { yellow: 'yellow' } }, key: takeKey() }], want: { raw: '', validated: { yellow: 'yellow' } }, wantInputRef: React.createRef(), key: takeKey() };
            examplesList.splice(destinationIndex, 0, newYellowExample);
            props.handleOnDrag(examplesList, props.tableIndex!, true);
        } else if (destinationIndex === examplesList.length) {
            newYellowExample = { inputs: [{ prog: { raw: '', validated: { yellow: 'yellow' } }, key: takeKey() }], want: { raw: '', validated: { yellow: 'yellow' } }, wantInputRef: React.createRef(), key: takeKey() };
            examplesList.splice(destinationIndex, 0, newYellowExample);
            const [reorderedExample] = examplesList.splice(sourceIndex, 1);
            examplesList.splice(destinationIndex, 0, reorderedExample);
            props.handleOnDrag(examplesList, props.tableIndex!, true);

        } else {
            const [reorderedExample] = examplesList.splice(sourceIndex, 1);
            examplesList.splice(destinationIndex, 0, reorderedExample);
            props.handleOnDrag(examplesList, props.tableIndex!);
        }
    }

    const reals = props.examples.map((example, i) => (
        <Draggable key={example.key} index={i} draggableId={example.key.toString()}>
            {(provided: any) => (
                <tr ref={provided.innerRef} {...provided.draggableProps} {...provided.dragHandleProps} tabIndex="-1">
                    <td>
                        <RemButton
                            onClick={props.disabled ? undefined : (() => remExample(example))}
                            title={'Remove this example'}
                        />
                    </td>
                    <Inputs
                        disabled={props.disabled}
                        globalEnv={props.globalEnv}
                        dummy={false}
                        inputs={example.inputs}
                        inputsChange={(inputs: InputArray) => exampleChange({ ...example, inputs },
                            example)}
                    />
                    <td>{/* empty cell to align with param dummy input */}</td>
                    <Outputs
                        globalEnv={props.globalEnv}
                        dummy={false}
                        formulas={props.formulas}
                        want={example.want}
                        row={i}
                    />
                    <td>{/* empty cell to align with top level formula dummy input */}</td>
                    <Want
                        disabled={props.disabled}
                        globalEnv={props.globalEnv}
                        dummy={false}
                        want={example.want}
                        wantInputRef={example.wantInputRef}
                        wantChange={(want: ProgramInput) => exampleChange({ ...example, want },
                            example)}
                    />
                </tr>
            )}
        </Draggable>
    ));

    const dummyRef: RefObject<HTMLTextAreaElement> = React.createRef();
    const dummy = (
        // index = examples.length
        <Draggable key={peekKey(props.paramNames.length)} index={props.examples.length} draggableId={peekKey().toString()}>
            {(provided: any) => (
                <tr ref={provided.innerRef} {...provided.draggableProps} {...provided.dragHandleProps} tabIndex="-1">
                    <td>{/* empty cell to offset rembutton */}</td>
                    <Inputs
                        disabled={props.disabled}
                        globalEnv={props.globalEnv}
                        dummy={true}
                        inputs={props.paramNames.map((_, i) => ({ prog: { raw: '', validated: { yellow: "yellow" } }, key: peekKey(i) }))}
                        inputsChange={(inputs: InputArray) => exampleChange({
                            inputs,
                            want: { raw: '', validated: { yellow: 'yellow' } },
                            wantInputRef: React.createRef(),
                            key: takeKey()
                        },
                            {})}
                    />
                    <td>{/* empty cell to align with param dummy input */}</td>
                    <Outputs
                        globalEnv={props.globalEnv}
                        dummy={true}
                        formulas={props.formulas}
                        want={{ raw: '', validated: { yellow: "yellow" } }}
                        row={0}
                    />
                    <td>{/* empty cell to align with top level formula dummy input */}</td>
                    <Want
                        disabled={props.disabled}
                        globalEnv={props.globalEnv}
                        dummy={true}
                        want={{ raw: '', validated: { yellow: 'yellow' } }}
                        wantInputRef={dummyRef}
                        wantChange={(want: ProgramInput) => exampleChange({
                            want,
                            inputs: props.paramNames.map((_) => ({ prog: { raw: '', validated: { yellow: 'yellow' } }, key: takeKey() })),
                            wantInputRef: dummyRef,
                            key: takeKey()
                        },
                            {})}
                    />
                </tr>
            )}
        </Draggable>
    );

    return (
        <DragDropContext onDragEnd={handleOnRowDrag}>
            <Droppable droppableId={"droppable-main-table" + props.tableIndex}>
                {(provided: any) => (
                    <tbody ref={provided.innerRef} {...provided.droppableProps}>
                        {[...reals, dummy]}
                        {provided.placeholder}
                    </tbody>
                )}
            </Droppable>
        </DragDropContext>
    );
}

export { SuccinctBody };