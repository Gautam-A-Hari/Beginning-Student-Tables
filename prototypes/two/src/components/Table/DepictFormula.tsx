import React from "react";
import { peekKey, takeKey, unparse_to_string } from "../../App";
import { RemButton } from "../RemButton";
import { BooleanFormula, Formula, isBooleanFormula, isFormulaNonEmpty, Yellow } from "../../input-definitions";
import { parseCheck } from "../../interpreter";
import { ValidatedArea } from "../ValidatedArea";

interface Props {

    disabled: boolean
    key: number
    formula: Formula
    depth: number
    numExamples: number

    formulaChange: (newForm: Formula) => void
}

function DepictFormula(props: Props) {
    function validProg(text:(string | undefined)) {
        try {
            parseCheck(text);
        } catch (e) {
            return false;
        }

        return true;
    }

    // this is pretty macabre...
    function remChild(deadChild: Formula) {
        if (isBooleanFormula(props.formula)) {
            const aliveChildren = props.formula.thenChildren.filter((child) => child !== deadChild);
        }
        const aliveChildren = isBooleanFormula(props.formula) ? props.formula.thenChildren.filter((child) => child !== deadChild) : [];
        props.formulaChange({ ...props.formula, thenChildren: aliveChildren });
    }

    function childChange(newChild: Formula, modChild: ({} | Formula)) {
        let children;
        if (isBooleanFormula(props.formula)) {
            if ((isFormulaNonEmpty(modChild) && props.formula.thenChildren.indexOf(modChild) === -1) || !isFormulaNonEmpty(modChild)) {
                children = [...props.formula.thenChildren, newChild];
            } else {
                children = props.formula.thenChildren.map((child) => child === modChild ? newChild : child);
            }
        }
        props.formulaChange({ ...props.formula, thenChildren: children });
    }

    // if the rawFormulaString doesn't parse, pass up the formula with the previous validated
    function validChildChange(newChild: Formula, modChild: ({} | Formula), newRawFormulaString:string) {
        try {
            childChange({
                ...newChild,
                prog: { raw: newRawFormulaString, validated: parseCheck(newRawFormulaString) }
            },
            newChild)
        } catch (e) {
            childChange({
                ...newChild,
                prog: { raw: newRawFormulaString, validated: newChild.prog.validated }
            },
            modChild)
        }
    }

    function countWidth(formula: Formula): number {
        if (!isBooleanFormula(formula)) {
            return 1;
        } else {
            return formula.thenChildren.reduce((acc, child) => acc + countWidth(child), 2);
        }
    }

    if (props.depth > 1) {
        return (
            <React.Fragment>
                <th>{/* empty cell to align with parent input */}</th>
                {isBooleanFormula(props.formula) ?
                    <React.Fragment>
                        {props.formula.thenChildren.map((child) => (
                            <DepictFormula
                                disabled={props.disabled}
                                key={child.key}
                                formula={child}
                                depth={props.depth - 1}
                                numExamples={props.numExamples}
                                formulaChange={(formula) => childChange(formula, child)}
                            />))}
                        <th>{/* empty cell to align with child input */}</th>
                    </React.Fragment>
                    : <script />}
            </React.Fragment>
        );
    } else {
        if (isBooleanFormula(props.formula)) {
            const reals = props.formula.thenChildren.map((child) => (
                <th key={child.key} colSpan={countWidth(child)} >
                    <div className='flex_horiz'>
                        <ValidatedArea
                            dummy={false}
                            placeholder={'Formula'}
                            text={props.disabled ? unparse_to_string(child.prog.validated) : undefined}
                            rawText={child.prog.raw}
                            disabled={props.disabled}
                            isValid={validProg}
                            onValid={(text) => validChildChange(child, child, text)}
                            onEmpty={() => childChange({
                                ...child,
                                prog: { raw: '', validated: {yellow: 'yellow' }}
                            },
                                child)}
                        />
                        <RemButton
                            title={'Remove this formula'}
                            onClick={props.disabled ? undefined : (() => remChild(child))}
                        />
                    </div>
                </th>
            ));

            const dummy = (
                <th key={peekKey()} colSpan={1}>
                    <div className='flex_horiz'>
                        <ValidatedArea
                            dummy={true}
                            placeholder='Formula'
                            text={props.disabled ? '' : undefined}
                            rawText=''
                            disabled={props.disabled}
                            isValid={validProg}
                            onValid={(text) => validChildChange({
                                prog: {raw: text, validated: {yellow: 'yellow'}},
                                outputs: Array(props.numExamples).fill({raw: '', validated: {yellow: 'yellow'}}),
                                key: takeKey()
                            }, {}, text)}
                            onEmpty={()=>null}
                        />
                    </div>
                </th>
            );

            return (
                <React.Fragment>
                    <th>{/* empty cell to align with parent input */}</th>
                    <React.Fragment>
                        {[...reals, dummy]}
                    </React.Fragment>
                </React.Fragment>
            );
        } else {

            return (
                <React.Fragment>
                    <th>{/* empty cell to align with parent input */}</th>
                </React.Fragment>
            );
        }
    }
}

export { DepictFormula };