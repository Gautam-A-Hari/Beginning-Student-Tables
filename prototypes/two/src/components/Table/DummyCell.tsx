import React from "react";
import { isRBOOL } from "../../global-definitions";
import { gray, pink, yellow } from "../../header";
import { isOutputNonYellow, isYellowProgramGray, Output } from "../../input-definitions";

interface Props {
    parentOutput: Output

}
function DummyCell(props:Props) {
    if ((!isOutputNonYellow(props.parentOutput) && isYellowProgramGray(props.parentOutput)) || (isOutputNonYellow(props.parentOutput) && isRBOOL(props.parentOutput) && props.parentOutput.value === false)) {
        return (
            <td className={'gray'}>
            </td>
        );
    } else if (props.parentOutput === pink || (isOutputNonYellow(props.parentOutput) && !isRBOOL(props.parentOutput))) {
        return (
            <td className={'pink'}>
            </td>
        );
    } else {
        return <td></td>;
    }
}

export { DummyCell };