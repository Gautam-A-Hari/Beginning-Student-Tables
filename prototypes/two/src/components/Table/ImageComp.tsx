import React, { useEffect, useState } from "react";
import { height, paint, width } from "../../image";
import { Image } from "../../image-definitions";

interface Props {
    image: Image
    key: number
}

export const ImageComp:React.FC<Props> = ({ image }) => {

    const [isOpen, changeIsOpen] = useState<boolean>(false);

    /**
     * changes isOpen state to false if the image Prop changes
     * this can happen when the input or the formula is changed
     */
    useEffect(() => {
        changeIsOpen(false);
    }, [image]);
    
    /**
     * handles the click on an image
     * if the image isZoomable, isOpen state is flipped
     * @param image the clicked on image
     */
    const handleViewClick = (image:Image) => {
        if (isZoomable(image)) {
            changeIsOpen(!isOpen);
        }
    }

    /**
     * determines whether the given image is larger than set dimensions
     * @param image the image to test
     * @returns boolean
     */
    const isZoomable = (image:Image) => {
        return (height(image) > 250 || width(image) > 250);
    }

    // TODO: isntead of a dialog pop-up, blur out everything behind enlarged image
    // if isOpen state is true, then dialog box is open with full size picture
    // else, image is rendered at scaled size
    return (
        <div className={isZoomable(image) ? 'image-output' : 'output'} onClick={() => handleViewClick(image)} >
                    {paint(image, true)}
                    {isOpen && (
                        <dialog
                            style={{ position: "absolute", cursor: "zoom-out" }}
                            open
                            onClick={() => handleViewClick(image)}
                        >
                            {paint(image)}
                        </dialog>
                    )}
        </div>
    )

}