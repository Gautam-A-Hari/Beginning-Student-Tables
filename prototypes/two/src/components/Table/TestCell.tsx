import Octicon, { Check } from "@primer/octicons-react";
import React from "react";
import { deepEquals, unparse } from "../../App";
import { ErrorMessage } from "../ErrorMessage";
import { Environment, Program } from "../../global-definitions";
import { yellow } from "../../header";
import { height, width } from "../../image";
import { isOutputNonYellow, isValidatedProgInputNonYellow, isYellowProgramGray, Output, ProgramInput } from "../../input-definitions";
import { interp, RIMAGE_T } from "../../interpreter";

interface Props {
    globalEnv : Environment
    output: Output
    want: ProgramInput
    
}

interface State {
    isOpen: boolean
}
class TestCell extends React.Component<Props, State> {
    /* have unparse(image) be a full size render, maybe by storing full size render in state? */
    constructor(props:Props) {
        super(props);
        this.handleViewClick = this.handleViewClick.bind(this);
        this.isZoomable = this.isZoomable.bind(this);
        this.state = {
            isOpen: false,
        }
    }
    // paint could be called with an extra param that specifies what type of render should be returned

    // Output ->
    handleViewClick(prog:Output) {
        // check if program returns an image
        if (!isOutputNonYellow(prog)) return false;
        if (this.isZoomable(prog)) {
            this.setState(
                { isOpen: !this.state.isOpen }
            );
        }
    }

    // Output -> Boolean
    // determines whether the image should be zoomable
    isZoomable(prog:Program) {
        if (isOutputNonYellow(prog)) {
            // maybe export Types not as type so we can have Types.RIMAGE_T
            return prog.type === RIMAGE_T ? (width(prog.value) > 250 || height(prog.value) > 250) : false;
        }
        return false;
    }

    componentDidUpdate(prevProps:Props) {
        if (prevProps.output !== this.props.output) {
            this.setState({
                isOpen: false
            });
        }
    }

    render() {
        let output = this.props.output;

        if (!isOutputNonYellow(output) && isYellowProgramGray(output)) {
            return (
                <td className={'gray'}>
                </td>
            );
        }

        if (!isOutputNonYellow(output) && "pink" in output) {
            return (
                <td className={'pink'}>
                </td>
            );
        }

        if (!isOutputNonYellow(output) && "yellow" in output) {
            return (
                <td className={'yellow'}>
                </td>
            );
        }

        if (output instanceof Error) {
            return <td><ErrorMessage error={output} /></td>
        }

        let want:ProgramInput;
        try {
            want = {raw: this.props.want.raw, validated: interp(this.props.want.validated, this.props.globalEnv)};
        } catch (e) {
            want = {raw: '', validated: yellow};
        }

        if (isValidatedProgInputNonYellow(want.validated) && isOutputNonYellow(output) && deepEquals(output, want.validated)) {
            return (
                <td className='output' onClick={() => this.handleViewClick(output)}>
                    {unparse(output, true)}
                    <div title={"Yay! It's right!"} className="check">
                        <Octicon
                            icon={Check} size="small" verticalAlign="middle" 
                            ariaLabel='Yay!' />
                    </div>
                    {this.state.isOpen && (
                        <dialog
                            style={{ position: "absolute" }}
                            open
                            onClick={() => this.handleViewClick(output)}
                        >
                            {unparse(output)}
                        </dialog>
                    )}
                </td>
            )
        } else if (isOutputNonYellow(output) && isValidatedProgInputNonYellow(output)){
            return (
                <td className='output' onClick={() => this.handleViewClick(output)}>
                    {this.state.isOpen && (
                        <dialog
                            style={{ position: "absolute" }}
                            open
                            onClick={() => this.handleViewClick(output)}
                        >
                            {unparse(output)}
                        </dialog>
                    )}
                    {unparse(output, true)}
                </td>
            );
        }
    }
}

export { TestCell };