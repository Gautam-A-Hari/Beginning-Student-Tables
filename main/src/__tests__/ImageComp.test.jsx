import { ImageComp } from '../components/Table/ImageComp';
import renderer from 'react-test-renderer';

/**
 * Snapshot test for image comp rendering
 */
it("renders correctly", () => {
    const imageComp = renderer.create(
        <ImageComp image={ {r: 150, mode: "solid", color: {'r':250, 'g':250, 'b':250, 'a':0}, type: "circle"} } key={1} />
    );
    let tree = imageComp.toJSON();
    expect(tree).toMatchSnapshot();

});