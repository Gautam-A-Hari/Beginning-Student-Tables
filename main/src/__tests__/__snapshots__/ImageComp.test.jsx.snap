// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="image-output"
  onClick={[Function]}
>
  <svg
    height={250}
    viewBox="0 0 250 250"
    width={250}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      style={
        Object {
          "transform": "scale(0.8333333333333334)",
        }
      }
    >
      <circle
        cx={150}
        cy={150}
        fill="rgba(250, 250, 250, 0)"
        r={150}
      />
    </g>
  </svg>
</div>
`;
